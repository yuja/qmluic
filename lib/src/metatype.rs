//! Qt metatypes.json data types.

use serde::{Deserialize, Serialize};

// TODO: Symbols are stored as owned strings, but we'll eventually want to intern them,
// where we can probably remove these temporary owned strings at all.

/// C++ access specifier.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum AccessSpecifier {
    Private,
    Protected,
    Public,
}

/// Revision number of method or property.
#[derive(
    Clone, Copy, Debug, Default, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize,
)]
#[repr(transparent)]
#[serde(transparent)]
pub struct Revision(pub u32);

impl Revision {
    pub fn is_zero(&self) -> bool {
        self.0 == 0
    }
}

/// Class metadata.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Class {
    pub class_name: String,
    pub qualified_class_name: String,
    #[serde(default)]
    pub super_classes: Vec<SuperClassSpecifier>,
    #[serde(default)]
    pub class_infos: Vec<ClassInfo>,
    #[serde(default)]
    pub interfaces: Vec<Vec<Interface>>, // no idea why this is nested array
    #[serde(default)]
    pub object: bool,
    #[serde(default)]
    pub gadget: bool,
    #[serde(default)]
    pub namespace: bool,

    #[serde(default)]
    pub enums: Vec<Enum>,
    #[serde(default)]
    pub properties: Vec<Property>,
    #[serde(default)]
    pub constructors: Vec<Method>, // only QObject appears to define this
    #[serde(default)]
    pub signals: Vec<Method>,
    #[serde(default)]
    pub slots: Vec<Method>,
    #[serde(default)]
    pub methods: Vec<Method>,
}

impl Class {
    /// Creates class metadata of object type.
    pub fn new<S>(name: S) -> Self
    where
        S: Into<String>,
    {
        let name: String = name.into();
        Class {
            class_name: unqualify_name(&name).to_owned(),
            qualified_class_name: name,
            object: true,
            ..Default::default()
        }
    }

    /// Creates class metadata of gadget type.
    pub fn new_gadget<S>(name: S) -> Self
    where
        S: Into<String>,
    {
        let name: String = name.into();
        Class {
            class_name: unqualify_name(&name).to_owned(),
            qualified_class_name: name,
            gadget: true,
            ..Default::default()
        }
    }

    /// Creates class metadata of object type with public super classes.
    pub fn with_supers<S, I>(name: S, supers: I) -> Self
    where
        S: Into<String>,
        I: IntoIterator,
        I::Item: Into<String>,
    {
        let name: String = name.into();
        let super_classes = supers
            .into_iter()
            .map(SuperClassSpecifier::public)
            .collect();
        Class {
            class_name: unqualify_name(&name).to_owned(),
            qualified_class_name: name,
            object: true,
            super_classes,
            ..Default::default()
        }
    }
}

fn unqualify_name(name: &str) -> &str {
    name.rsplit_once("::").map(|(_, n)| n).unwrap_or(name)
}

/// Super class reference with the access specifier.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SuperClassSpecifier {
    pub name: String,
    pub access: AccessSpecifier,
}

impl SuperClassSpecifier {
    /// Creates public super class specifier.
    pub fn public<S>(name: S) -> Self
    where
        S: Into<String>,
    {
        SuperClassSpecifier {
            name: name.into(),
            access: AccessSpecifier::Public,
        }
    }
}

/// Extra class metadata. (e.g. `""QML.Element"`" name)
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ClassInfo {
    pub name: String,
    pub value: String,
}

impl ClassInfo {
    /// Creates class info pair.
    pub fn new<S, T>(name: S, value: T) -> Self
    where
        S: Into<String>,
        T: Into<String>,
    {
        ClassInfo {
            name: name.into(),
            value: value.into(),
        }
    }
}

/// Qt plugin interface identifier. (see `Q_DECLARE_INTERFACE()`)
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Interface {
    pub class_name: String,
    pub id: String,
}

/// Enum (and flag) metadata.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Enum {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alias: Option<String>,
    pub is_class: bool,
    pub is_flag: bool,
    pub values: Vec<String>,
}

impl Enum {
    /// Creates enum metadata.
    pub fn new<S>(name: S) -> Self
    where
        S: Into<String>,
    {
        Enum {
            name: name.into(),
            ..Default::default()
        }
    }

    /// Creates flag metadata.
    pub fn new_flag<S, T>(name: S, alias: T) -> Self
    where
        S: Into<String>,
        T: Into<String>,
    {
        Enum {
            name: name.into(),
            alias: Some(alias.into()),
            is_flag: true,
            ..Default::default()
        }
    }

    /// Creates enum metadata with values.
    pub fn with_values<S, I>(name: S, values: I) -> Self
    where
        S: Into<String>,
        I: IntoIterator,
        I::Item: Into<String>,
    {
        Enum {
            name: name.into(),
            values: values.into_iter().map(|n| n.into()).collect(),
            ..Default::default()
        }
    }
}

/// Property metadata.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Property {
    pub name: String,
    pub r#type: String,

    /// See `Q_PRIVATE_PROPERTY()`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub private_class: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub member: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub read: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub write: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reset: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub notify: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bindable: Option<String>,

    #[serde(default, skip_serializing_if = "Revision::is_zero")]
    pub revision: Revision,
    pub designable: bool,
    pub scriptable: bool,
    pub stored: bool,
    pub user: bool,
    pub constant: bool,
    pub r#final: bool,
    pub required: bool,

    /// Property index in the current meta object. (new in Qt 6)
    pub index: Option<i32>,
}

impl Property {
    /// Creates property metadata.
    pub fn new<S, T>(name: S, type_name: T) -> Self
    where
        S: Into<String>,
        T: Into<String>,
    {
        Property {
            name: name.into(),
            r#type: type_name.into(),
            ..Default::default()
        }
    }

    /// Creates property metadata of final.
    pub fn new_final<S, T>(name: S, type_name: T) -> Self
    where
        S: Into<String>,
        T: Into<String>,
    {
        Property {
            name: name.into(),
            r#type: type_name.into(),
            r#final: true,
            ..Default::default()
        }
    }
}

impl Default for Property {
    fn default() -> Self {
        Self {
            name: String::new(),
            r#type: String::new(),
            private_class: None,
            member: None,
            read: None,
            write: None,
            reset: None,
            notify: None,
            bindable: None,
            revision: Revision(0),
            designable: true,
            scriptable: true,
            stored: true,
            user: false,
            constant: false,
            r#final: false,
            required: false,
            index: None,
        }
    }
}

/// Signal, slot, or callable function metadata.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Method {
    pub name: String,
    pub access: AccessSpecifier,
    pub return_type: String,
    #[serde(default)]
    pub arguments: Vec<Argument>,
    #[serde(default, skip_serializing_if = "Revision::is_zero")]
    pub revision: Revision,
}

impl Default for Method {
    fn default() -> Self {
        Self {
            name: String::new(),
            access: AccessSpecifier::Public,
            return_type: String::new(),
            arguments: vec![],
            revision: Revision(0),
        }
    }
}

/// Signal, slot, or callable function argument.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Argument {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub r#type: String,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CompilationUnit {
    #[serde(default)]
    pub classes: Vec<Class>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_file: Option<String>,
    #[serde(default)]
    pub output_revision: i32,
}

/// Collects all classes from metatypes.json document.
///
/// The JSON document should consist of either a compilation unit object `{"classes": ...}`
/// or an array of such objects `[{"classes": ...}, ...]`.
pub fn extract_classes_from_str(json_data: &str) -> serde_json::Result<Vec<Class>> {
    // Don't use untagged enum to dispatch [{...}] and {...}. Doing that would move any
    // parsing error to the document level and line/column information would be lost.
    if json_data.trim_start().starts_with('{') {
        let unit: CompilationUnit = serde_json::from_str(json_data)?;
        Ok(unit.classes)
    } else {
        let units: Vec<CompilationUnit> = serde_json::from_str(json_data)?;
        Ok(units.into_iter().flat_map(|u| u.classes).collect())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize_class_decl() {
        let data = r###"{
            "className": "QAction",
            "enums": [
                {
                    "isClass": false,
                    "isFlag": false,
                    "name": "Priority",
                    "values": [
                        "LowPriority",
                        "NormalPriority",
                        "HighPriority"
                    ]
                }
            ],
            "object": true,
            "properties": [
                {
                    "constant": false,
                    "designable": true,
                    "final": false,
                    "name": "checkable",
                    "notify": "changed",
                    "read": "isCheckable",
                    "required": false,
                    "scriptable": true,
                    "stored": true,
                    "type": "bool",
                    "user": false,
                    "write": "setCheckable"
                }
            ],
            "qualifiedClassName": "QAction",
            "signals": [
                {
                    "access": "public",
                    "name": "changed",
                    "returnType": "void"
                },
                {
                    "access": "public",
                    "arguments": [
                        {
                            "name": "checked",
                            "type": "bool"
                        }
                    ],
                    "name": "triggered",
                    "returnType": "void"
                }
            ],
            "slots": [
                {
                    "access": "public",
                    "name": "trigger",
                    "returnType": "void"
                },
                {
                    "access": "public",
                    "name": "hover",
                    "returnType": "void"
                },
                {
                    "access": "public",
                    "arguments": [
                        {
                            "type": "bool"
                        }
                    ],
                    "name": "setChecked",
                    "returnType": "void"
                }
            ],
            "superClasses": [
                {
                    "access": "public",
                    "name": "QObject"
                }
            ]
        }"###;
        let _meta: Class = serde_json::from_str(data).unwrap();
    }

    #[test]
    fn extract_classes_unit() {
        let data = r###"{
            "classes": [
                {"className": "Foo", "qualifiedClassName": "Foo", "object": true},
                {"className": "Bar", "qualifiedClassName": "Bar", "object": true}
            ]
        }"###;
        let classes = extract_classes_from_str(data).unwrap();
        assert_eq!(classes.len(), 2);
    }

    #[test]
    fn extract_classes_nested() {
        let data = r###"[
            {"classes": [{"className": "Foo", "qualifiedClassName": "Foo", "object": true}]},
            {"classes": [{"className": "Bar", "qualifiedClassName": "Bar", "object": true}]}
        ]"###;
        let classes = extract_classes_from_str(data).unwrap();
        assert_eq!(classes.len(), 2);
    }
}
