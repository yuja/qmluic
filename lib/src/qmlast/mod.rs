//! QML parser interface and document model.

use std::error::Error;
use std::fmt;
use std::fs;
use std::io;
use std::ops::Range;
use std::path::Path;
use tree_sitter::{Parser, Query, QueryCursor, Tree};

mod astutil;
mod expr;
mod object;
mod term;

pub use self::expr::*; // re-export
pub use self::object::*; // re-export
pub use self::term::*; // re-export
pub use tree_sitter::Node; // re-export

/// Object holding QML source text and parsed tree.
#[derive(Clone, Debug)]
pub struct UiDocument {
    source: String,
    tree: Tree,
}

impl UiDocument {
    /// Creates parsed tree for the given QML source.
    ///
    /// The parsing doesn't fail even if the QML source has a syntax error. Instead, a node
    /// representing the error is inserted.
    pub fn with_source(source: String) -> Self {
        let mut parser = new_parser();
        let tree = parser
            .parse(source.as_bytes(), None)
            .expect("no timeout nor cancellation should have been made");
        UiDocument { source, tree }
    }

    /// Creates parsed tree from the given QML file.
    ///
    /// The parsing doesn't fail even if the QML source has a syntax error. Instead, a node
    /// representing the error is inserted.
    pub fn read(path: impl AsRef<Path>) -> io::Result<Self> {
        let source = fs::read_to_string(path)?;
        Ok(Self::with_source(source))
    }

    pub fn has_syntax_error(&self) -> bool {
        self.tree.root_node().has_error()
    }

    /// Collects syntax errors from the parsed tree.
    pub fn collect_syntax_errors<'a, B>(&'a self) -> B
    where
        B: FromIterator<ParseError<'a>>,
    {
        let query =
            Query::new(self.tree.language(), "(ERROR) @a").expect("static query must be valid");
        let mut cursor = QueryCursor::new();
        let matches = cursor.matches(&query, self.tree.root_node(), self.source.as_bytes());
        matches
            .map(|m| ParseError::new(m.captures[0].node, ParseErrorKind::InvalidSyntax))
            .collect()
    }

    pub fn source(&self) -> &str {
        &self.source
    }

    /// Root node of the parsed tree.
    pub fn root_node(&self) -> Node {
        self.tree.root_node()
    }
}

fn new_parser() -> Parser {
    let mut parser = Parser::new();
    parser
        .set_language(tree_sitter_qmljs::language())
        .expect("QML grammar should be compatible with parser");
    parser
}

/// Semantic or syntax error occurred while parsing QML source.
#[derive(Clone, Debug)]
pub struct ParseError<'tree> {
    node: Node<'tree>,
    kind: ParseErrorKind,
}

/// Details of QML parse error.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ParseErrorKind {
    InvalidSyntax,
    UnexpectedNodeKind,
    MissingField(&'static str),
    DuplicatedBinding,
    DuplicatedObjectId,
}

impl<'tree> ParseError<'tree> {
    pub fn new(node: Node<'tree>, kind: ParseErrorKind) -> Self {
        ParseError { node, kind }
    }

    pub fn node(&self) -> Node<'tree> {
        self.node
    }

    pub fn kind(&self) -> ParseErrorKind {
        self.kind
    }

    pub fn start_byte(&self) -> usize {
        self.node.start_byte()
    }

    pub fn end_byte(&self) -> usize {
        self.node.end_byte()
    }

    pub fn byte_range(&self) -> Range<usize> {
        self.node.byte_range()
    }
}

impl fmt::Display for ParseError<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use ParseErrorKind::*;
        match self.kind {
            InvalidSyntax => write!(f, "syntax error"),
            UnexpectedNodeKind => write!(f, "unexpected node kind: {}", self.node.kind()),
            MissingField(name) => write!(f, "missing field: {}", name),
            DuplicatedBinding => write!(f, "duplicated binding"),
            DuplicatedObjectId => write!(f, "duplicated object id"),
        }
    }
}

impl Error for ParseError<'_> {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn syntax_error() {
        let doc = UiDocument::with_source(
            r###"
            import
            "###
            .to_owned(),
        );
        assert!(doc.has_syntax_error());
        let errors: Vec<_> = doc.collect_syntax_errors();
        assert_eq!(errors.len(), 1);
    }
}
