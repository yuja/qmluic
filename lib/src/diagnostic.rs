//! Utility for error reporting.

use crate::qmlast::ParseError;
use camino::{Utf8Path, Utf8PathBuf};
use std::ops::Range;
use std::slice;

/// Type (or level) of diagnostic message.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum DiagnosticKind {
    Error,
}

/// Diagnostic message.
#[derive(Clone, Debug)]
pub struct Diagnostic {
    kind: DiagnosticKind,
    byte_range: Range<usize>,
    message: String,
    labels: Vec<(Range<usize>, String)>,
}

impl Diagnostic {
    /// Creates new diagnostic message of the given `kind`.
    pub fn new<S>(kind: DiagnosticKind, byte_range: Range<usize>, message: S) -> Self
    where
        S: Into<String>,
    {
        Diagnostic {
            kind,
            byte_range,
            message: message.into(),
            labels: Vec::new(),
        }
    }

    /// Creates new error message.
    pub fn error<S>(byte_range: Range<usize>, message: S) -> Self
    where
        S: Into<String>,
    {
        Self::new(DiagnosticKind::Error, byte_range, message)
    }

    /// Builds diagnostic with the given label attached.
    pub fn with_label<S>(mut self, byte_range: Range<usize>, message: S) -> Self
    where
        S: Into<String>,
    {
        self.push_label(byte_range, message);
        self
    }

    /// Builds diagnostic with the given labels attached.
    pub fn with_labels<I, S>(mut self, labels: I) -> Self
    where
        I: IntoIterator<Item = (Range<usize>, S)>,
        S: Into<String>,
    {
        self.extend_labels(labels);
        self
    }

    pub fn kind(&self) -> DiagnosticKind {
        self.kind
    }

    pub fn start_byte(&self) -> usize {
        self.byte_range.start
    }

    pub fn end_byte(&self) -> usize {
        self.byte_range.end
    }

    pub fn byte_range(&self) -> Range<usize> {
        Range {
            start: self.byte_range.start,
            end: self.byte_range.end,
        }
    }

    pub fn message(&self) -> &str {
        &self.message
    }

    pub fn labels(&self) -> &[(Range<usize>, String)] {
        &self.labels
    }

    /// Add a label to this diagnostic.
    pub fn push_label<S>(&mut self, byte_range: Range<usize>, message: S)
    where
        S: Into<String>,
    {
        self.labels.push((byte_range, message.into()));
    }

    /// Add labels to this diagnostic.
    pub fn extend_labels<I, S>(&mut self, labels: I)
    where
        I: IntoIterator<Item = (Range<usize>, S)>,
        S: Into<String>,
    {
        self.labels
            .extend(labels.into_iter().map(|(r, s)| (r, s.into())));
    }
}

impl From<&ParseError<'_>> for Diagnostic {
    fn from(error: &ParseError) -> Self {
        Self::error(error.byte_range(), error.to_string())
    }
}

impl From<ParseError<'_>> for Diagnostic {
    fn from(error: ParseError) -> Self {
        Self::from(&error)
    }
}

/// Manages diagnostic messages.
#[derive(Clone, Debug, Default)]
pub struct Diagnostics {
    diagnostics: Vec<Diagnostic>,
}

impl Diagnostics {
    pub fn new() -> Self {
        Diagnostics::default()
    }

    pub fn is_empty(&self) -> bool {
        self.diagnostics.is_empty()
    }

    pub fn len(&self) -> usize {
        self.diagnostics.len()
    }

    pub fn iter(&self) -> slice::Iter<'_, Diagnostic> {
        self.diagnostics.iter()
    }

    pub fn push<T>(&mut self, diag: T)
    where
        T: Into<Diagnostic>,
    {
        self.diagnostics.push(diag.into())
    }

    /// Extracts error from the given `result` and pushes it. Returns the success value if any.
    pub fn consume_err<T, E>(&mut self, result: Result<T, E>) -> Option<T>
    where
        E: Into<Diagnostic>,
    {
        match result {
            Ok(x) => Some(x),
            Err(e) => {
                self.push(e);
                None
            }
        }
    }
}

impl<'a> IntoIterator for &'a Diagnostics {
    type Item = &'a Diagnostic;
    type IntoIter = slice::Iter<'a, Diagnostic>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

impl<T> Extend<T> for Diagnostics
where
    T: Into<Diagnostic>,
{
    fn extend<I>(&mut self, iter: I)
    where
        I: IntoIterator<Item = T>,
    {
        self.diagnostics.extend(iter.into_iter().map(|d| d.into()))
    }
}

/// Manages per-file store of diagnostic messages.
#[derive(Clone, Debug, Default)]
pub struct ProjectDiagnostics {
    file_diagnostics: Vec<(Utf8PathBuf, Diagnostics)>, // in insertion order
}

impl ProjectDiagnostics {
    pub fn new() -> Self {
        ProjectDiagnostics::default()
    }

    pub fn is_empty(&self) -> bool {
        // push() guarantees that each file store is not empty
        self.file_diagnostics.is_empty()
    }

    pub fn iter(&self) -> slice::Iter<'_, (Utf8PathBuf, Diagnostics)> {
        self.file_diagnostics.iter()
    }

    pub fn diagnostics(&self) -> impl Iterator<Item = (&Utf8Path, &Diagnostic)> {
        self.file_diagnostics
            .iter()
            .flat_map(|(p, ds)| ds.iter().map(|d| (p.as_ref(), d)))
    }

    pub fn push<P>(&mut self, path: P, diagnostics: Diagnostics)
    where
        P: AsRef<Utf8Path>,
    {
        if !diagnostics.is_empty() {
            self.file_diagnostics
                .push((path.as_ref().to_owned(), diagnostics));
        }
    }
}
