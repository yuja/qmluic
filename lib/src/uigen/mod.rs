//! Qt user interface XML (.ui) generator.

use crate::diagnostic::{Diagnostic, Diagnostics};
use crate::qmlast::{UiImportSource, UiProgram};
use crate::qmldir;
use crate::qmldoc::UiDocument;
use crate::typemap::{Class, Module, ModuleData, ModuleId, Type, TypeMap, TypeSpace};
use thiserror::Error;

mod expr;
mod gadget;
mod layout;
mod object;
mod property;
mod xmlutil;

pub use self::expr::*; // re-export
pub use self::gadget::*; // re-export
pub use self::layout::*; // re-export
pub use self::object::*; // re-export

pub type XmlError = quick_xml::Error;
pub type XmlResult<T> = quick_xml::Result<T>;
pub type XmlWriter<W> = quick_xml::Writer<W>;

/// Builds `UiForm` from the given `doc`.
pub fn build(
    base_ctx: &BuildContext,
    doc: &UiDocument,
    diagnostics: &mut Diagnostics,
) -> Option<UiForm> {
    let program = diagnostics.consume_err(UiProgram::from_node(doc.root_node(), doc.source()))?;
    // TODO: do not build unnamed module space if directory module available
    let module_data = make_doc_module_data(doc, &program, base_ctx.type_map, diagnostics);
    let type_space = get_doc_type(base_ctx.type_map, doc)
        .unwrap_or_else(|| Type::Module(Module::new(&module_data, base_ctx.type_map)));
    let ctx = BuildDocContext::new(doc, type_space, base_ctx);
    let (obj, cls) =
        object::resolve_object_definition(&ctx, program.root_object_node(), diagnostics)?;
    let root_object =
        UiObject::from_object_definition(&ctx, &cls, &obj, ContainerKind::Any, diagnostics)?;
    Some(UiForm {
        class: doc.type_name().map(|s| s.to_owned()),
        root_object,
    })
}

fn get_doc_type<'a>(type_map: &'a TypeMap, doc: &UiDocument) -> Option<Type<'a>> {
    if let Some(doc_path) = doc.path() {
        let base_dir =
            qmldir::normalize_path(doc_path.parent().expect("doc path should point to file"));
        type_map
            .get_module(ModuleId::Directory(base_dir.into()))
            .and_then(|m| {
                m.get_type(
                    doc.type_name()
                        .expect("doc loaded from file should have type"),
                )
            })
    } else {
        None
    }
}

fn make_doc_module_data(
    doc: &UiDocument,
    program: &UiProgram,
    type_map: &TypeMap,
    diagnostics: &mut Diagnostics,
) -> ModuleData {
    let mut module_data = ModuleData::with_builtins();
    for imp in program.imports() {
        if imp.alias().is_some() {
            diagnostics.push(Diagnostic::error(
                imp.node().byte_range(),
                "aliased import is not supported",
            ));
            continue;
        }
        // TODO: warn that version field is ignored
        // TODO: anchor diagnostic message onto imp.source() node
        let id = match imp.source() {
            UiImportSource::Identifier(x) => {
                ModuleId::Named(x.to_string(doc.source()).into_owned().into())
            }
            UiImportSource::String(x) => {
                if let Some(p) = doc.path().and_then(|p| p.parent()) {
                    ModuleId::Directory(qmldir::normalize_path(p.join(&x)).into())
                } else {
                    diagnostics.push(Diagnostic::error(
                        imp.node().byte_range(),
                        "cannot resolve directory path against inline QML document",
                    ));
                    continue;
                }
            }
        };
        if type_map.contains_module(&id) {
            module_data.import_module(id);
        } else {
            diagnostics.push(Diagnostic::error(
                imp.node().byte_range(),
                "module not found",
            ));
        }
    }
    module_data
}

/// Resources passed around the UI object constructors.
#[derive(Clone, Debug)]
pub struct BuildContext<'a> {
    type_map: &'a TypeMap,
    action_class: Class<'a>,
    form_layout_class: Class<'a>,
    grid_layout_class: Class<'a>,
    hbox_layout_class: Class<'a>,
    layout_class: Class<'a>,
    layout_attached_class: Class<'a>,
    push_button_class: Class<'a>,
    spacer_item_class: Class<'a>,
    tab_widget_class: Class<'a>,
    tab_widget_attached_class: Class<'a>,
    vbox_layout_class: Class<'a>,
    widget_class: Class<'a>,
}

#[derive(Clone, Debug)]
struct BuildDocContext<'a, 's> {
    source: &'s str,
    action_class: Class<'a>,
    form_layout_class: Class<'a>,
    grid_layout_class: Class<'a>,
    hbox_layout_class: Class<'a>,
    layout_class: Class<'a>,
    layout_attached_class: Class<'a>,
    push_button_class: Class<'a>,
    spacer_item_class: Class<'a>,
    tab_widget_class: Class<'a>,
    tab_widget_attached_class: Class<'a>,
    vbox_layout_class: Class<'a>,
    widget_class: Class<'a>,
    type_space: Type<'a>,
}

impl<'a> BuildContext<'a> {
    /// Sets up context with the given `type_map`.
    pub fn prepare(type_map: &'a TypeMap) -> Result<Self, BuildContextError> {
        const MODULE_NAME: &str = "qmluic.QtWidgets";
        let module = type_map
            .get_module(&ModuleId::Named(MODULE_NAME.into()))
            .ok_or(BuildContextError::ModuleNotFound(MODULE_NAME))?;
        let get_class = |name| {
            if let Some(Type::Class(cls)) = module.get_type(name) {
                Ok(cls)
            } else {
                Err(BuildContextError::ClassNotFound(name))
            }
        };
        Ok(BuildContext {
            type_map,
            action_class: get_class("QAction")?,
            form_layout_class: get_class("QFormLayout")?,
            grid_layout_class: get_class("QGridLayout")?,
            hbox_layout_class: get_class("QHBoxLayout")?,
            layout_class: get_class("QLayout")?,
            layout_attached_class: get_class("QLayoutAttached")?,
            push_button_class: get_class("QPushButton")?,
            spacer_item_class: get_class("QSpacerItem")?,
            tab_widget_class: get_class("QTabWidget")?,
            tab_widget_attached_class: get_class("QTabWidgetAttached")?,
            vbox_layout_class: get_class("QVBoxLayout")?,
            widget_class: get_class("QWidget")?,
        })
    }
}

impl<'a, 's> BuildDocContext<'a, 's> {
    fn new(doc: &'s UiDocument, type_space: Type<'a>, base_ctx: &BuildContext<'a>) -> Self {
        BuildDocContext {
            source: doc.source(),
            action_class: base_ctx.action_class.clone(),
            form_layout_class: base_ctx.form_layout_class.clone(),
            grid_layout_class: base_ctx.grid_layout_class.clone(),
            hbox_layout_class: base_ctx.hbox_layout_class.clone(),
            layout_class: base_ctx.layout_class.clone(),
            layout_attached_class: base_ctx.layout_attached_class.clone(),
            push_button_class: base_ctx.push_button_class.clone(),
            spacer_item_class: base_ctx.spacer_item_class.clone(),
            tab_widget_class: base_ctx.tab_widget_class.clone(),
            tab_widget_attached_class: base_ctx.tab_widget_attached_class.clone(),
            vbox_layout_class: base_ctx.vbox_layout_class.clone(),
            widget_class: base_ctx.widget_class.clone(),
            type_space,
        }
    }
}

/// Error occurred while setting up [`BuildContext`].
#[derive(Clone, Debug, Error)]
pub enum BuildContextError {
    #[error("required class not found: {0}")]
    ClassNotFound(&'static str),
    #[error("required module not found: {0}")]
    ModuleNotFound(&'static str),
}
