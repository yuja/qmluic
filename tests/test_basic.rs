use qmluic::diagnostic::Diagnostics;
use qmluic::metatype;
use qmluic::qmlast::UiDocument;
use qmluic::typemap::TypeMap;
use qmluic::uigen::XmlWriter;
use qmluic_cli::UiBuilder;
use std::fs;
use std::path::Path;

#[test]
fn test_translate_example() {
    insta::assert_snapshot!(translate_file("examples/SettingsDialog.qml",));
    insta::assert_snapshot!(translate_file("examples/MainWindow.qml"));
    insta::assert_snapshot!(translate_file("examples/VariousLayouts.qml"));
}

fn translate_file(path: impl AsRef<Path>) -> String {
    let mut type_map = TypeMap::with_primitive_types();
    type_map.extend(load_metatypes());
    let doc = UiDocument::read(path).unwrap();
    let mut diagnostics = Diagnostics::new();
    let form = UiBuilder::new(&type_map, &doc, &mut diagnostics)
        .build()
        .unwrap();
    let mut buf = Vec::new();
    form.serialize_to_xml(&mut XmlWriter::new_with_indent(&mut buf, b' ', 1))
        .unwrap();
    String::from_utf8(buf).unwrap()
}

fn load_metatypes() -> Vec<metatype::Class> {
    let paths = [
        "contrib/metatypes/qmluic_metatypes.json",
        "contrib/metatypes/qt5core_metatypes.json",
        "contrib/metatypes/qt5gui_metatypes.json",
        "contrib/metatypes/qt5widgets_metatypes.json",
    ];
    paths
        .iter()
        .flat_map(|p| {
            let data = fs::read_to_string(p).unwrap();
            metatype::extract_classes_from_str(&data).unwrap()
        })
        .collect()
}
