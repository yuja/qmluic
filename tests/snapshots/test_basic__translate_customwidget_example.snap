---
source: tests/test_basic.rs
expression: "common::translate_file(base.join(\"MainDialog.qml\")).unwrap()"
---
<ui version="4.0">
 <class>MainDialog</class>
 <widget class="QDialog">
  <layout class="QVBoxLayout">
   <item>
    <widget class="SettingsForm" name="settingsForm">
    </widget>
   </item>
   <item>
    <widget class="MyDialogButtonBox" name="buttonBox">
    </widget>
   </item>
  </layout>
 </widget>
</ui>

